### PR TITLE
test(ivy): example of locality working with component selectors

### PR DESCRIPTION
**Pull request by mhevery**
This spec demonstrates how component can respect locality in a backwards compatible way which
does not break tree shaking.

Locality is important goal to the compiler because it allows tho compiler to translate each of
the Angular types (`@Component`, `@Directive`, `@Pipe`, `@NgModule`) in a way which does not
require any information outside of what is directly available in the decorator. This allows
source to be compiled in an independent way. The consequence of this is simple compiler mental
model, fast compiles, and publishing compiled code to NPM.

Tree shaking is important because we want to ensure that only symbols which are directly needed
by the components are retained.

The proposal solves two problems:
1. How to compile `@Component` without knowing what other components, directives or pipes are
   active.
2. Produce output which is tree shakable.

Goal 1 is achieved by compiling `@Component` templates as if there were no other components,
directives or pipes present. The resulting `ComponentDef` object than allows 'patching' of
component, directives, and pipes at a later point in time (before the first render). On first
render the Ivy render engine resolves the component, directives, and pipes. Subsequent renders
are fast since the resolution has happened. The "patching" process is achieved by the
`@NgModule`'s `defineNgModule` function. While this allows the code to run without any further
tool chain it breaks tree shaking.

Goal 2 is achieved by allowing build-optimizer to rearrange the `defineNgModule` calls such
that the information is moved to the specific components and the `defineNgModule` is removed.
This transformation allows the subsequent build tools to remove any unreferenced symbols.

The moving of definitions from `defineNgModule` to `defineComponent` needs to be done in a way
which takes selectors into account. (Just because `@NgModule` declares component `Foo` does not
mean that it is used in an application. Blindly applying `Foo` to all components would retain
it even if no one used it.) For this reason the generated code contains pragmas for the
build-optimizer which tell it how the code should be rearranged in tree-shakable way. The
important thing to realize is that the rearranging has no semantic impact, it is purely an
optimization operation to allow tree-shaker to do its job.

When the build-optimizer looks to redistribute component, directive, and pipe references it
needs to know if the destination needs it. To do that build-optimizer needs to be able to
compare pipe names as well as execute selectors to see if any of the directives match a
destination. For this reason the generated code has pragmas in comments which communicate
what pipe names are needed and which elements and attributes are in template to see if
any of the potential directives match it. The pipe or directive is only added if the
build-optimizer can determine that there has been a match.

### CODE REVIEW

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+ * ```
+ *
+ *
+ * `@__NG_MODULE_DECLARATIONS__`
+ *
+ * Communicates that `@NgModule` definition follows. The job of the build-optimizer is to remove
+ * all instances of `@__NG_MODULE_DECLARATIONS__` by inlining the corresponding component, directive
+ * and pipe references into the concrete components which need the information. For inline process
+ * to be tree-shakable it must take usage of the destination component into account. (Does the
+ * component uses particular particular component, directive or pipe).
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Typo: particular particular -> a particular

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+ */
+
+@Pipe({name: 'noop'})
+class MyPipe implements PipeTransform {
+  transform(value: any, ...args: any[]) { return value; }
+  // NORMATIVE
+  /**
+   * @__NG_DECLARE_PIPE__: 'foo'
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Should this be `noop` ?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+    factory: function MyPipe_Factory() { return new MyPipe(); },
+  });
+  // /NORMATIVE
+}
+
+@Directive({selector: '[foo]'})
+class FooDirective {
+  // NORMATIVE
+  /**
+   * @__NG_DECLARE_DIRECTIVE__: 'foo'
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Should this be `['', foo']`?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+     * - `MyPipe` - gets inlined because its name matches a name in the
+     *   `@__NG_PIPE_TARGETS__`.
+     * - `UnusedDirective` - does not get inline because its selector does not match
+     *   `@__NG_DIRECTIVE_SELECTOR_TARGETS__`
+     * - `FooDirective` - gets inlined because its selector matches at least one element in
+     *   `@__NG_DIRECTIVE_SELECTOR_TARGETS__`.
+     *
+     * The result is that build-optimizer will add the following two lines into the
+     * `defineComponent`.
+     * NOTE: the closure is need since moving references around could refer to a symbol before it is
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Typo: need -> needed

(Same on line 237)

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+})
+class MyApp {
+  // NORMATIVE
+  /**
+   * @__NG_DECLARE_DIRECTIVE__: ['my-app']
+   * @__NG_DIRECTIVE_SELECTOR_TARGETS__: [ ['my-comp'] ]
+   */
+  static ngComponentDef = $r3$.ɵdefineComponent({
+    type: MyComponent,
+    tag: 'my-cmp',
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Should be `MyApp` and `'my-app'`, right?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+interface NgModuleDef {
+  patchImports: (destination: Type<any>) => void
+}
+
+/**
+ * Creates selector map definition object.
+ *
+ * `NgModule` through `declarations`, `imports` and `exports` can determine which component,
+ * directive and pipes can be used in which component templates. Selector map definition
+ * provides that runtime information. The information is formated in such a way so that global
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Typo: formated -> formatted

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+ * optimizer can rearrange the referees so that the code becomes available for tree-shaking.
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Typo: referees -> references

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+ *
+ * @param declarations
+ * @param patch
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Looks like this JSDoc is outdated?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+    }
+  };
+}
+
+function addDirectiveOrPipe(destination: Type<any>, directiveOrPipeDependency: Type<any>): void {
+  const ngComponentDef: ComponentDef<any> = (destination as any).ngComponentDef;
+  if (ngComponentDef) {
+    const ngComponentDefRef: ComponentDef<any> = (directiveOrPipeDependency as any).ngComponentDef;
+    const ngDirectiveDefRef: DirectiveDef<any> = (directiveOrPipeDependency as any).ngDirectiveDef;
+    if (ngComponentDef || ngDirectiveDefRef) {
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: I think you can short-circuit one of the megamorphic property accesses here if you define the `directiveDef` and `componentRef` in the same line (so the second won't be accessed unless the first isn't found). 

```ts
const defRef = (directiveOrPipeDependency as any).ngComponentRef || 
(directiveOrPipeDependency as any).ngDirectiveDef;

if (defRef) {
  ngComponentDef.directiveDefs = (...).push(defRef);
}
```

**Review by kara — COMMENTED** [review]
Nits and typos. +1 on the plan itself.

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+import {ComponentDef, DirectiveDef} from '../../../src/render3';
+import {PipeDef} from '../../../src/render3/interfaces/definition';
+
+type $boolean$ = boolean;
+type $TRUST_TYPE$ = any;
+type $MyComponent$ = MyComponent;
+type $MyApp$ = MyApp;
+
+/**
+ * This spec demonstrates how component can respect locality in a backwards compatible way which
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: fixed

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+ * does not break tree shaking.
+ *
+ * Locality is important goal to the compiler because it allows tho compiler to translate each of
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: fixed

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+ * Tree shaking is important because we want to ensure that only symbols which are directly needed
+ * by the components are retained.
+ *
+ * The proposal solves two problems:
+ * 1. How to compile `@Component` without knowing what other components, directives or pipes are
+ *    active.
+ * 2. Produce output which is tree shakable.
+ *
+ * Goal 1 is achieved by compiling `@Component` templates as if there were no other components,
+ * directives or pipes present. The resulting `ComponentDef` object than allows 'patching' of
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: fixed

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+ * component, directives, and pipes at a later point in time (before the first render). On first
+ * render the Ivy render engine resolves the component, directives, and pipes. Subsequent renders
+ * are fast since the resolution has happened. The "patching" process is achieved by the
+ * `@NgModule`'s `defineNgModule` function. While this allows the code to run without any further
+ * tool chain it breaks tree shaking.
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: added

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+ *
+ * The moving of definitions from `defineNgModule` to `defineComponent` needs to be done in a way
+ * which takes selectors into account. (Just because `@NgModule` declares component `Foo` does not
+ * mean that it is used in an application. Blindly applying `Foo` to all components would retain
+ * it even if no one used it.) For this reason the generated code contains pragmas for the
+ * build-optimizer which tell it how the code should be rearranged in tree-shakable way. The
+ * important thing to realize is that the rearranging has no semantic impact, it is purely an
+ * optimization operation to allow tree-shaker to do its job.
+ *
+ * When the build-optimizer looks to redistribute component, directive, and pipe references it
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: added

File: packages/core/src/render3/interfaces/definition.ts [hunk 15]
```diff
@@ -163,6 +163,23 @@ export interface ComponentDef<T> extends DirectiveDef<T> {
    * children only.
    */
   readonly viewProviders?: Provider[];
+
+  /**
+   * List of directives which are active on this component.
+   *
+   * The property is either an array of `DirectiveDef`s or a function which returns the array of
+   * `DireciveDef`s. The function is necessary to be able to support forward declarations.
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/render3/interfaces/definition.ts]: fixed

File: packages/core/test/render3/compiler_canonical/local_selectors.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: fixed

File: packages/core/test/render3/compiler_canonical/local_selectors.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: yes, changed

File: packages/core/test/render3/compiler_canonical/local_selectors.ts (hunk 3 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: yes, fixed

File: packages/core/test/render3/compiler_canonical/local_selectors.ts (hunk 4 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: fixed

File: packages/core/test/render3/compiler_canonical/local_selectors.ts (hunk 5 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: yes, good catch

File: packages/core/test/render3/compiler_canonical/local_selectors.ts (hunk 6 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: fixed

File: packages/core/test/render3/compiler_canonical/local_selectors.ts (hunk 7 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: fixed

File: packages/core/test/render3/compiler_canonical/local_selectors.ts (hunk 8 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: fixed

File: packages/core/test/render3/compiler_canonical/local_selectors.ts (hunk 9 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical/local_selectors.ts]: fixed

File: packages/core/src/render3/interfaces/definition.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (kara) [reply packages/core/src/render3/interfaces/definition.ts]: Looks like the typo's still here? (Line 165)

File: packages/core/src/render3/interfaces/definition.ts [hunk 16]
```diff
@@ -157,6 +157,23 @@ export interface ComponentDef<T> extends DirectiveDef<T> {
...
+   */
+  directiveDefs: (() => Array<DirectiveDef<any>|ComponentDef<any>>)|
+      Array<DirectiveDef<any>|ComponentDef<any>>|null;
```
Comment (kara) [inline packages/core/src/render3/interfaces/definition.ts]: Seems like we need to add `directiveDefs` to `ComponentDefArgs` interface as well, right? If we are passing it into `defineComponent`.

File: packages/core/src/render3/definition.ts [hunk 17]
```diff
@@ -56,7 +56,9 @@ export function defineComponent<T>(componentDefinition: ComponentDefArgs<T>): Co
     afterViewChecked: type.prototype.ngAfterViewChecked || null,
     onDestroy: type.prototype.ngOnDestroy || null,
     onPush: (componentDefinition as ComponentDefArgs<T>).changeDetection ===
-        ChangeDetectionStrategy.OnPush
+        ChangeDetectionStrategy.OnPush,
+    directiveDefs: null,
```
Comment (kara) [inline packages/core/src/render3/definition.ts]: Wouldn't this need to catch any directive defs passed into `defineComponent`? That's where build-optimizer will add it, no?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 18]
```diff
@@ -0,0 +1,382 @@
...
+    factory: function MyPipe_Factory() { return new MyPipe(); },
+  });
+  // /NORMATIVE
+}
+
+@Directive({selector: '[foo]'})
+class FooDirective {
+  // NORMATIVE
+  /**
+   * @__NG_DECLARE_DIRECTIVE__: ['', foo']
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Missing open quote?  `foo'` -> `'foo'`

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 19]
```diff
@@ -0,0 +1,382 @@
...
+     *   `@__NG_DIRECTIVE_SELECTOR_TARGETS__`
+     * - `MyComponent` - gets inlined because its selector matches at least one element in
+     *   `@__NG_DIRECTIVE_SELECTOR_TARGETS__`.
+     *
+     * The result is that build-optimizer will add the following line into the `defineComponent`.
+     *
+     * NOTE: the closure is need since moving references around could refer to a symbol before it is
+     * initialized.
+     */
+    // directiveDefs: () => [MyComponent],
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Should this be `MyComponent.ngComponentDef`?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 20]
```diff
@@ -0,0 +1,382 @@
...
+     * - `FooDirective` - gets inlined because its selector matches at least one element in
+     *   `@__NG_DIRECTIVE_SELECTOR_TARGETS__`.
+     *
+     * The result is that build-optimizer will add the following two lines into the
+     * `defineComponent`.
+     * NOTE: the closure is needed since moving references around could refer to a symbol before it
+     * is
+     * initialized.
+     */
+    // directiveDefs: () => [FooDirective],
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Should this be `FooDirective.ngDirectiveDef`?

File: packages/core/src/render3/interfaces/definition.ts (hunk 16 above)
Comment (trotyl) [inline packages/core/src/render3/interfaces/definition.ts]: Wouldn't `DirectiveDef<any>|ComponentDef<any>` same as `DirectiveDef<any>` due to the [inheritance](https://github.com/angular/angular/blob/5ab9d4d437ab27fa8df248012af895204136ca8e/packages/core/src/render3/interfaces/definition.ts#L124)?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 21]
```diff
@@ -0,0 +1,382 @@
...
+ * by the components are retained.
+ *
+ * The proposal solves two problems:
+ * 1. How to compile `@Component` without knowing what other components, directives or pipes are
+ *    active.
+ * 2. Produce output which is tree shakable.
+ *
+ * Goal 1 is achieved by compiling `@Component` templates as if there were no other components,
+ * directives or pipes present. The resulting `ComponentDef` object then allows 'patching' of
+ * component, directives, and pipes at a later point in time (before the first render). On first
```
Comment (trotyl) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: `component` -> `components` for plural consistency? Another case in the line below.

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 22]
```diff
@@ -0,0 +1,382 @@
...
+ * any of the potential directives match it. The pipe or directive is only added if the
+ * build-optimizer can determine that there has been a match.
+ */
+
+/**
+ * build-optimizer pragmas description
+ *
+ * `@__NG_DECLARE_PIPE__` followed by pipe name.
+ *
+ * Communicates that pipe declaration of a give name follows.
```
Comment (trotyl) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: `a give name` -> `a given name`?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 23]
```diff
@@ -0,0 +1,382 @@
...
+ *
+ *
+ * `@__NG_DECLARE_DIRECTIVE__` (`@@__NG_DECLARE_COMPONENT__`) followed by parsed selector.
```
Comment (trotyl) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: `@@__NG_DECLARE_COMPONENT__` why double `@` symbol here?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 24]
```diff
@@ -0,0 +1,382 @@
...
+ *
+ * Communicates that a directive (or component) declaration with a given selector follows.
+ *
+ * The selector format is `[elementName, attr1, value1, attr2, value2, ...]` This format
```
Comment (trotyl) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: How does this handle comma combinator like `selector: 'my-button, button[myButton]'`?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 25]
```diff
@@ -0,0 +1,382 @@
...
+ * Communicates a parsed version of all of the element and attributes present in the component's
+ * view. The build-optimizer can use this information to determine if a given directive (or
+ * component) selector matches this list. If match is determined that the candidate directive is
+ * added to this component definition.
+ *
+ * The format of array of array of targets is:
+ * ```
+ * [
+ *   [ elementName1, attr1_1, value1_1, attr1_2, attr1_2, ...],
+ *   [ elementName2, attr2_1, value2_1, attr2_2, attr2_2, ...],
```
Comment (trotyl) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: `attr1_2, attr1_2` -> `attr1_2, value1_2`? As well as the next line?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 26]
```diff
@@ -0,0 +1,382 @@
...
+ * ```
+ *
+ *
+ * `@__NG_MODULE_DECLARATIONS__`
+ *
+ * Communicates that `@NgModule` definition follows. The job of the build-optimizer is to remove
+ * all instances of `@__NG_MODULE_DECLARATIONS__` by inlining the corresponding component, directive
+ * and pipe references into the concrete components which need the information. For inline process
+ * to be tree-shakable it must take usage of the destination component into account. (Does the
+ * component uses a particular component, directive or pipe).
```
Comment (trotyl) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: `uses` -> `use` in favor of `Does`?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 27]
```diff
@@ -0,0 +1,382 @@
...
+     * NON-NORMATIVE COMMENT Demonstrates what build-optimizer will add.
+     *
+     * In this case `MyComponent` is part of `MyLibModule` which has
+     * `declarations: [MyComponent, MyPipe, UnusedDirective, FooDirective]`.
+     *
+     * - `MyComponent` - does not get inlined because its selector does not match
+     *   `@__NG_DIRECTIVE_SELECTOR_TARGETS__`
+     * - `MyPipe` - gets inlined because its name matches a name in the
+     *   `@__NG_PIPE_TARGETS__`.
+     * - `UnusedDirective` - does not get inline because its selector does not match
```
Comment (trotyl) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: `get inline` -> `get inlined`?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 28]
```diff
@@ -0,0 +1,382 @@
...
+        $r3$.ɵe();
+        $r3$.ɵPp(2, 'noop' as $TRUST_TYPE$);
+      }
+      $r3$.ɵt(1, $r3$.ɵb($r3$.ɵpb1(2, ctx.name)));
+    }
+  });
+  // /NORMATIVE
+}
+// NORMATIVE
+const $MyComponent_attrs_0$ = ['foo', ''];
```
Comment (trotyl) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: TDZ error for `const`?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 29]
```diff
@@ -0,0 +1,382 @@
...
+}
+
+////////////////////////////////////////////
+// TODO: move to `src` folder
+////////////////////////////////////////////
+
+interface NgModuleType<T> extends Type<T> {
+  ngModuleDef: NgModuleDef;
+}
+interface NgModuleDef {
```
Comment (trotyl) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: Shouldn't `NgModuleDef` be generic like other `Def`s?

File: packages/core/test/render3/compiler_canonical/local_selectors.ts [hunk 30]
```diff
@@ -0,0 +1,382 @@
...
+  patchImports: (destination: Type<any>) => void;
+}
+
+/**
+ * Creates selector map definition object.
+ *
+ * `NgModule` through `declarations`, `imports` and `exports` can determine which component,
+ * directive and pipes can be used in which component templates. Selector map definition
```
Comment (trotyl) [inline packages/core/test/render3/compiler_canonical/local_selectors.ts]: `which component templates` -> `which component template`?

**Pull request closed without merging**